lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+        }
+
+        $checksum = $algorithmSplit[1]
+        return @{ 'URL' = $urlSplit[0];
+                  'CHECKSUM' = $checksum;
+                  'HASHING_ALGORITHM' = $hashingAlgorithm }
+    }
+
+    return @{ 'URL' = $url;
+              'CHECKSUM' = $null;
+              'HASHING_ALGORITHM' = $null }
+}
+
+
+function Start-DownloadFile {
+    Param(
+        [Parameter(Mandatory=$True)]
+        [System.Uri]$Uri,
+        [string]$OutFile,
+        [switch]$SkipIntegrityCheck=$false
+    )
+
+    if(!$OutFile) {
+        $OutFile = $Uri.PathAndQuery.Substring($Uri.PathAndQuery.LastIndexOf("/") + 1)
+        if(!$OutFile) {
+            throw "The ""OutFile"" parameter needs to be specified"
+        }
+    }
+
+    $webClient = New-Object System.Net.WebClient
+    Start-ExecuteWithRetry { $webClient.DownloadFile($Uri, $OutFile) }
+
+    if(!$SkipIntegrityCheck) {
+        $fragment = $Uri.Fragment.Trim('#')
+        if (!$fragment){
+            return
+        }
+        $details = $fragment.Split("=")
+        $algorithm = $details[0]
+        $hash = $details[1]
+        if($algorithm -in @("SHA1", "SHA256", "SHA384", "SHA512", "MACTripleDES", "MD5", "RIPEMD160")){
+            Test-FileIntegrity -File $OutFile -Algorithm $algorithm -ExpectedHash $hash
+        } else {
+            Throw "Hash algorithm $algorithm not recognized."
+        }
+    }
+}
+
+
+# Returns the full path of the package after it is downloaded using
+# the URL parameter (a checksum may optionally be specified). The
+# package is cached on the disk until the installation successfully finishes.
+# If the hook fails, on the second run this function will return the cached
+# package path if checksum is given and it matches.
+function Get-PackagePath {
+    Param(
+        [Parameter(Mandatory=$True)]
+        [string]$URL,
+        [string]$Checksum="",
+        [string]$HashingAlgorithm=""
+    )
+
+    $packagePath = Join-Path $env:TEMP $URL.Split('/')[-1]
+    if (Test-Path $packagePath) {
+        if ($Checksum -and $HashingAlgorithm) {
+            if (Test-FileIntegrity -File $packagePath -Algorithm $HashingAlgorithm -ExpectedHash $Checksum) {
+                return $packagePath
+            }
+        }
+        Remove-Item -Recurse -Force -Path $packagePath
+    }
+
+    if ($Checksum -and $HashingAlgorithm) {
+        Start-DownloadFile -Uri "$URL#$HashingAlgorithm=$Checksum" -OutFile $packagePath
+    } else {
+        Start-DownloadFile -Uri $URL -OutFile $packagePath -SkipIntegrityCheck
+    }
+
+    return $packagePath
+}
+
+
+# Installs a package after it is downloaded from the Internet and checked for
+# integrity with SHA1 checksum. Accepts as parameters: an URL, an optional
+# 'Checksum' with its 'HashingAlgorithm' and 'ArgumentList' which can be passed
+# if the installer requires unattended installation.
+# Supported packages formats are: '.exe' and '.msi'
+function Install-Package {
```

Comment:
there is an ```Install-Msi``` function in ```JujuWindowsUtils```. While A generic function that can install both MSI and exe is not terrible, I would advise you create individual functions that installs your dependencies.

For example, one function to install git, one function to install vcredist and so on. Some dependencies may have special considerations, or dependencies of their own. As the list of dependencies grows or changes, it may become difficult to treat edge cases.

Small functions that install individual packages, decouples the logic of installing one package from the rest of your code. No need for if conditions or special cases for one package or another. Here is an example of what I mean:

https://github.com/cloudbase/nova-hyperv-charm/blob/master/lib/Modules/OVSCharmUtils/OVSCharmUtils.psm1#L313